In [2]:
import numpy as np 
import pandas as pd

In [3]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv']))



In [4]:
# 158 with wc 2022 teams

In [5]:
df.shape

(199, 17)

In [6]:
df

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0   0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1   0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2   0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3   0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4   0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   
..       ...       ...        ...           ...       ...           ...   
2   0.293846  0.280000   0.040769      0.140769  0.298462      0.099231   
3   0.553077  0.411538  -0.104615      0.026923 -0.352308     -0.238462   
4   0.239231  0.424615   0.068462      0.406923  0.001538      0.000769   
5  -0.093077 -0.067692   0.272308      0.108462  0.023077      0.133846   
6   0.572308  0.006154  -0.037692      0.076154 -0.416923     -0.153846   

    takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0    0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1    0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2    0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3    0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4    0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   
..        ...         ...        ...              ...           ...       ...   
2    0.011538    0.217692   0.535385         0.263846     -0.331538 -0.016154   
3   -0.254615   -0.268462  -0.125385         0.029231     -0.119231 -0.279231   
4   -0.250769   -0.051538   0.008462         0.157692     -0.395385 -0.520000   
5   -0.214615    0.561538   0.419231         0.098462      0.150000 -0.136923   
6   -0.170769   -0.191538  -0.466154        -0.107692      0.344615 -0.072308   

    aerials_z   fouls_z  fouled_z    nsxg_z  results  
0   -0.020000 -0.224615  0.100769 -0.124615        2  
1    0.196923 -0.016923  0.033846  0.146923        1  
2    0.683846  0.213846  0.353846  0.786923        9  
3    0.038462  0.039231  0.022308  0.692308        5  
4    0.572308 -0.016154 -0.096923  0.890769        5  
..        ...       ...       ...       ...      ...  
2   -0.445385  0.037692  0.296923  0.340769       17  
3   -0.150769 -0.266154 -0.406154 -0.210000       18  
4    0.040769  0.096154  0.170769  0.409231       22  
5   -0.103846  0.186154 -0.163077 -0.001538       24  
6    0.683077  0.365385 -0.138462 -0.017692        8  

[199 rows x 17 columns]

In [7]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.607556   0.074575      0.303607  0.176107   
xg_z             0.607556  1.000000   0.401252      0.663849  0.334035   
crosses_z        0.074575  0.401252   1.000000      0.647671  0.238391   
boxtouches_z     0.303607  0.663849   0.647671      1.000000  0.382879   
passes_z         0.176107  0.334035   0.238391      0.382879  1.000000   
progpasses_z     0.036467  0.119222   0.246861      0.224493  0.757784   
takeons_z        0.249140  0.336452   0.144901      0.218674  0.369406   
progruns_z       0.188312  0.407663   0.303640      0.425804  0.600402   
tackles_z        0.255029  0.385452   0.216750      0.407687  0.752082   
interceptions_z  0.199810  0.309643   0.195412      0.317672  0.750774   
clearances_z    -0.198859 -0.402970  -0.222664     -0.411559 -0.142323   
blocks_z         0.002386  0.037704   0.030626      0.146541  0.394555   
aerials_z        0.065199  0.041713   0.237783      0.245014 -0.019196   
fouls_z         -0.139404 -0.090299   0.017189     -0.070728  0.025848   
fouled_z         0.033498  0.156988   0.188782      0.163910  0.302025   
nsxg_z           0.366439  0.659697   0.661716      0.857421  0.504541   
results         -0.611941 -0.310435   0.021751     -0.204948 -0.143200   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.036467   0.249140    0.188312   0.255029   
xg_z                 0.119222   0.336452    0.407663   0.385452   
crosses_z            0.246861   0.144901    0.303640   0.216750   
boxtouches_z         0.224493   0.218674    0.425804   0.407687   
passes_z             0.757784   0.369406    0.600402   0.752082   
progpasses_z         1.000000   0.143839    0.349855   0.529591   
takeons_z            0.143839   1.000000    0.536898   0.346489   
progruns_z           0.349855   0.536898    1.000000   0.537707   
tackles_z            0.529591   0.346489    0.537707   1.000000   
interceptions_z      0.561454   0.341074    0.426120   0.709325   
clearances_z        -0.007591  -0.099967   -0.074454  -0.221618   
blocks_z             0.370701   0.182073    0.288658   0.272809   
aerials_z            0.033879  -0.086779   -0.022921   0.056041   
fouls_z              0.183029   0.057999    0.012751  -0.038256   
fouled_z             0.181446   0.288267    0.248663   0.284206   
nsxg_z               0.355076   0.387291    0.529295   0.492715   
results             -0.059694  -0.150401   -0.194182  -0.194861   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.199810     -0.198859  0.002386   0.065199 -0.139404   
xg_z                    0.309643     -0.402970  0.037704   0.041713 -0.090299   
crosses_z               0.195412     -0.222664  0.030626   0.237783  0.017189   
boxtouches_z            0.317672     -0.411559  0.146541   0.245014 -0.070728   
passes_z                0.750774     -0.142323  0.394555  -0.019196  0.025848   
progpasses_z            0.561454     -0.007591  0.370701   0.033879  0.183029   
takeons_z               0.341074     -0.099967  0.182073  -0.086779  0.057999   
progruns_z              0.426120     -0.074454  0.288658  -0.022921  0.012751   
tackles_z               0.709325     -0.221618  0.272809   0.056041 -0.038256   
interceptions_z         1.000000     -0.127883  0.305893   0.012275  0.078056   
clearances_z           -0.127883      1.000000  0.243153   0.254803  0.253985   
blocks_z                0.305893      0.243153  1.000000   0.043416  0.201614   
aerials_z               0.012275      0.254803  0.043416   1.000000  0.087295   
fouls_z                 0.078056      0.253985  0.201614   0.087295  1.000000   
fouled_z                0.317386      0.074323  0.125723   0.027949  0.234694   
nsxg_z                  0.417540     -0.403278  0.172225   0.141886 -0.019848   
results                -0.126919      0.074508 -0.079578  -0.054570  0.128954   

                 fo

In [23]:
df.shape

(192, 17)

In [25]:
# df.to_csv('WC-stats.csv', index=False)


In [82]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [83]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [84]:
y = df['results']

In [85]:
X = df.drop(columns=['results'])

In [86]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [87]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.42629810615686115

In [88]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [90]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 1.282 +/- 0.149
    progruns_z0.096 +/- 0.029
    xg_z    0.095 +/- 0.030
    takeons_z0.026 +/- 0.013
neg_mean_absolute_percentage_error
    goals_z 0.988 +/- 0.115
neg_mean_squared_error
    goals_z 108.981 +/- 12.672
    progruns_z8.139 +/- 2.494
    xg_z    8.047 +/- 2.509
    takeons_z2.195 +/- 1.067


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.